sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -153,6 +153,16 @@ object SetOperationPushDown extends Rule[LogicalPlan] with PredicateHelper {
         )
       )
 
+    // Push down limit into union
+    case Limit(exp, Union(left, right), optimized) if !optimized =>
```

Comment:
This works, but doesn't look good to me. Actually I don't have a good idea for it either, maybe we should just add comments to explain the assumption here.

cc @marmbrus @yhuai